lib/picos_select/picos_select.ml

```diff
@@ -373,6 +374,21 @@ let[@alert "-handler"] cancel_after computation ~seconds exn_bt =
   if not (Computation.try_attach computation remover) then
     Trigger.signal remover
 
+let[@alert "-handler"] timeout ~seconds =
+  let request outer to_result =
+    let inner =
+      Computation.with_action to_result outer @@ fun _ to_result outer ->
+      Computation.return outer to_result
+    in
+    let canceler =
+      Trigger.from_action () inner @@ fun _ _ inner ->
+      Computation.cancel inner exit_exn_bt
+    in
+    if Computation.try_attach outer canceler then
+      cancel_after inner ~seconds exit_exn_bt
+  in
+  Event.from_request { request }
```

Comment:
This is somewhat tricky as we basically convert from cancelation to returning a value and we are also careful to cancel the request in case it is no longer needed.